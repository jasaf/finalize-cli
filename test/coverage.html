
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<title>cmd: Go Coverage Report</title>
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">github.com/jasaf/finalize-cli/cmd/main.go (0.0%)</option>
				
				<option value="file1">github.com/jasaf/finalize-cli/internal/console/console.go (68.0%)</option>
				
				<option value="file2">github.com/jasaf/finalize-cli/internal/file/file.go (94.4%)</option>
				
				<option value="file3">github.com/jasaf/finalize-cli/internal/helpers/dateformat.go (100.0%)</option>
				
				<option value="file4">github.com/jasaf/finalize-cli/internal/item/item.go (100.0%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">package main

import (
        "flag"
        "log"
        "os"

        "github.com/jasaf/finalize-cli/internal/console"
        "github.com/jasaf/finalize-cli/internal/file"
        "github.com/jasaf/finalize-cli/internal/item"
)

func main() <span class="cov0" title="0">{
        fs := flag.CommandLine
        userInput, err := console.ParseUserInput(fs, os.Args[1:])
        if err != nil </span><span class="cov0" title="0">{
                log.Fatalf("Error while parsing user input, %s\n", err)
        }</span>

        <span class="cov0" title="0">items, err := file.ReadCsvFile(userInput.FileName)
        if err != nil </span><span class="cov0" title="0">{
                log.Fatalf("Error while reading csv file, %s\n", err)
        }</span>

        <span class="cov0" title="0">categorisedItems := item.CategoriseItems(items, userInput.FromDate, userInput.ToDate)

        if err := console.PrintJson(categorisedItems); err != nil </span><span class="cov0" title="0">{
                log.Fatalf("Error while printing json")
        }</span>
}
</pre>
		
		<pre class="file" id="file1" style="display: none">package console

import (
        "encoding/json"
        "errors"
        "flag"
        "fmt"
        "strings"

        "github.com/jasaf/finalize-cli/internal/helpers"
        "github.com/jasaf/finalize-cli/internal/types"
)

func ParseUserInput(fs *flag.FlagSet, args []string) (*types.UserInput, error) <span class="cov8" title="1">{
        var inputFlag string
        var filterDateFlag string

        fs.StringVar(&amp;inputFlag, "input", "", "Insert CSV file name with extension")
        fs.StringVar(&amp;filterDateFlag, "date", "", "Insert date range with the following format [from:to]")

        if err := fs.Parse(args); err != nil </span><span class="cov0" title="0">{
                return nil, err
        }</span>

        <span class="cov8" title="1">if len(inputFlag) &lt; 1 </span><span class="cov8" title="1">{
                return nil, errors.New("file name input flag is empty")
        }</span>
        <span class="cov8" title="1">if len(filterDateFlag) &lt; 1 </span><span class="cov8" title="1">{
                return nil, errors.New("filter date flag is empty")
        }</span>

        <span class="cov8" title="1">betweenDates := strings.Split(filterDateFlag, ":")

        if len(betweenDates) &lt; 2 </span><span class="cov8" title="1">{
                return nil, errors.New("could not find the dates separator")
        }</span>

        <span class="cov8" title="1">fromDate, err := helpers.StringToDate(betweenDates[0])
        if err != nil </span><span class="cov0" title="0">{
                return nil, err
        }</span>

        <span class="cov8" title="1">toDate, err := helpers.StringToDate(betweenDates[1])
        if err != nil </span><span class="cov0" title="0">{
                return nil, err
        }</span>

        <span class="cov8" title="1">return &amp;types.UserInput{
                FileName: inputFlag,
                FromDate: fromDate,
                ToDate:   toDate,
        }, nil</span>
}

func PrintJson(data any) error <span class="cov0" title="0">{
        jsonData, err := json.MarshalIndent(data, "", "  ")
        if err != nil </span><span class="cov0" title="0">{
                return err
        }</span>

        <span class="cov0" title="0">fmt.Println(string(jsonData))

        return nil</span>
}
</pre>
		
		<pre class="file" id="file2" style="display: none">package file

import (
        "bytes"
        "encoding/csv"
        "os"
        "strconv"

        "github.com/jasaf/finalize-cli/internal/helpers"
        "github.com/jasaf/finalize-cli/internal/types"
)

func ReadCsvFile(name string) (*[]types.RowItem, error) <span class="cov8" title="1">{
        data, err := os.ReadFile(name)
        if err != nil </span><span class="cov8" title="1">{
                return nil, err
        }</span>

        <span class="cov8" title="1">r := csv.NewReader(bytes.NewReader(data))

        records, err := r.ReadAll()
        if err != nil </span><span class="cov0" title="0">{
                return nil, err
        }</span>

        <span class="cov8" title="1">var items []types.RowItem
        for _, record := range records[1:] </span><span class="cov8" title="1">{
                amount, err := strconv.ParseFloat(record[3], 64)
                if err != nil </span><span class="cov8" title="1">{
                        return nil, err
                }</span>

                <span class="cov8" title="1">date, err := helpers.StringToDate(record[0])
                if err != nil </span><span class="cov8" title="1">{
                        return nil, err
                }</span>

                <span class="cov8" title="1">newItem := types.RowItem{
                        Date:        date,
                        Category:    record[1],
                        Subcategory: record[2],
                        Amount:      amount,
                        Currency:    record[4],
                        Region:      record[5],
                }

                items = append(items, newItem)</span>
        }

        <span class="cov8" title="1">return &amp;items, nil</span>
}
</pre>
		
		<pre class="file" id="file3" style="display: none">package helpers

import (
        "time"
)

const (
        defaultFormat = "2006-01-02"
)

func StringToDate(date string) (time.Time, error) <span class="cov8" title="1">{
        parsedDate, err := time.Parse(defaultFormat, date)

        return parsedDate, err
}</span>
</pre>
		
		<pre class="file" id="file4" style="display: none">package item

import (
        "time"

        "github.com/jasaf/finalize-cli/internal/types"
)

func CategoriseItems(items *[]types.RowItem, fromDate, toDate time.Time) *types.CategoryTotals <span class="cov8" title="1">{
        filteredItems := make(types.CategoryTotals)

        for _, record := range *items </span><span class="cov8" title="1">{
                if record.Date.Before(fromDate) || record.Date.After(toDate) </span><span class="cov8" title="1">{
                        continue</span>
                }

                <span class="cov8" title="1">_, categoryExists := filteredItems[record.Category]
                if !categoryExists </span><span class="cov8" title="1">{
                        filteredItems[record.Category] = make(types.SubCategoryWithCurrencyTotal)
                }</span>

                <span class="cov8" title="1">categoryMap := filteredItems[record.Category]
                _, subCategoryExists := categoryMap[record.Subcategory]
                if !subCategoryExists </span><span class="cov8" title="1">{
                        categoryMap[record.Subcategory] = make(types.CurrencyTotal)
                }</span>

                <span class="cov8" title="1">currencyTotal := categoryMap[record.Subcategory]
                _, currencyExists := currencyTotal[record.Currency]
                if !currencyExists </span><span class="cov8" title="1">{
                        currencyTotal[record.Currency] = 0
                }</span>

                <span class="cov8" title="1">currencyTotal[record.Currency] += record.Amount</span>
        }

        <span class="cov8" title="1">return &amp;filteredItems</span>
}
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
