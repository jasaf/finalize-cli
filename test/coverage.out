mode: set
github.com/jasaf/finalize-cli/cmd/main.go:13.13,16.16 3 0
github.com/jasaf/finalize-cli/cmd/main.go:16.16,18.3 1 0
github.com/jasaf/finalize-cli/cmd/main.go:20.2,21.16 2 0
github.com/jasaf/finalize-cli/cmd/main.go:21.16,23.3 1 0
github.com/jasaf/finalize-cli/cmd/main.go:25.2,27.60 2 0
github.com/jasaf/finalize-cli/cmd/main.go:27.60,29.3 1 0
github.com/jasaf/finalize-cli/internal/item/item.go:9.96,12.32 2 1
github.com/jasaf/finalize-cli/internal/item/item.go:12.32,13.64 1 1
github.com/jasaf/finalize-cli/internal/item/item.go:13.64,14.12 1 1
github.com/jasaf/finalize-cli/internal/item/item.go:17.3,18.22 2 1
github.com/jasaf/finalize-cli/internal/item/item.go:18.22,20.4 1 1
github.com/jasaf/finalize-cli/internal/item/item.go:22.3,24.25 3 1
github.com/jasaf/finalize-cli/internal/item/item.go:24.25,26.4 1 1
github.com/jasaf/finalize-cli/internal/item/item.go:28.3,30.22 3 1
github.com/jasaf/finalize-cli/internal/item/item.go:30.22,32.4 1 1
github.com/jasaf/finalize-cli/internal/item/item.go:34.3,34.50 1 1
github.com/jasaf/finalize-cli/internal/item/item.go:37.2,37.23 1 1
github.com/jasaf/finalize-cli/internal/file/file.go:13.57,15.16 2 1
github.com/jasaf/finalize-cli/internal/file/file.go:15.16,17.3 1 1
github.com/jasaf/finalize-cli/internal/file/file.go:19.2,22.16 3 1
github.com/jasaf/finalize-cli/internal/file/file.go:22.16,24.3 1 0
github.com/jasaf/finalize-cli/internal/file/file.go:26.2,27.37 2 1
github.com/jasaf/finalize-cli/internal/file/file.go:27.37,29.17 2 1
github.com/jasaf/finalize-cli/internal/file/file.go:29.17,31.4 1 1
github.com/jasaf/finalize-cli/internal/file/file.go:33.3,34.17 2 1
github.com/jasaf/finalize-cli/internal/file/file.go:34.17,36.4 1 1
github.com/jasaf/finalize-cli/internal/file/file.go:38.3,47.33 2 1
github.com/jasaf/finalize-cli/internal/file/file.go:50.2,50.20 1 1
github.com/jasaf/finalize-cli/internal/helpers/dateformat.go:11.51,15.2 2 1
github.com/jasaf/finalize-cli/internal/console/console.go:14.80,21.39 5 1
github.com/jasaf/finalize-cli/internal/console/console.go:21.39,23.3 1 0
github.com/jasaf/finalize-cli/internal/console/console.go:25.2,25.24 1 1
github.com/jasaf/finalize-cli/internal/console/console.go:25.24,27.3 1 1
github.com/jasaf/finalize-cli/internal/console/console.go:28.2,28.29 1 1
github.com/jasaf/finalize-cli/internal/console/console.go:28.29,30.3 1 1
github.com/jasaf/finalize-cli/internal/console/console.go:32.2,34.27 2 1
github.com/jasaf/finalize-cli/internal/console/console.go:34.27,36.3 1 1
github.com/jasaf/finalize-cli/internal/console/console.go:38.2,39.16 2 1
github.com/jasaf/finalize-cli/internal/console/console.go:39.16,41.3 1 0
github.com/jasaf/finalize-cli/internal/console/console.go:43.2,44.16 2 1
github.com/jasaf/finalize-cli/internal/console/console.go:44.16,46.3 1 0
github.com/jasaf/finalize-cli/internal/console/console.go:48.2,52.8 1 1
github.com/jasaf/finalize-cli/internal/console/console.go:55.32,57.16 2 0
github.com/jasaf/finalize-cli/internal/console/console.go:57.16,59.3 1 0
github.com/jasaf/finalize-cli/internal/console/console.go:61.2,63.12 2 0
